### ISSUE TITLE
very serious pgn parsing bug

### ISSUE THREAD

**Issue by eguneys** [niklasf/chessops#172]
This guy is parsed wrong

https://lichess.org/study/nk2t0m1n/NnOi2xkd

[Event "East Orange"]
[Site "https://lichess.org/study/nk2t0m1n/NnOi2xkd"]
[Date "1957.??.??"]
[White "Fischer, R."]
[Black "Sherwin, J."]
[Result "1-0"]
[Annotator "Fritz 5.00 (90s)"]
[Variant "Standard"]
[ECO "B40"]
[Opening "Sicilian Defense: French Variation"]

1. e4 c5 2. Nf3 e6 3. d3 Nc6 4. g3 Nf6 { out of book } 5. Bg2 Be7 6. O-O O-O 7. Nbd2 Rb8 8. Re1 d6 { Prevents intrusion on e5 } 9. c3 { Consolidates b4+d4 } 9... b6 10. d4 Qc7 11. e5 { This push gains space. White gets in control } 11... Nd5 $16 (11... dxe5!? { and Black can hope to survive } 12. Nxe5 Nxe5 13. dxe5 Nd5) 12. exd6 $14 (12. c4 Ndb4 13. exd6 Qxd6 $16) 12... Bxd6 13. Ne4 c4 14. Nxd6 Qxd6 15. Ng5 Nce7 16. Qc2 { Threatening mate... how? } 16... Ng6 17. h4 Nf6 (17... Rd8 18. Nxh7 Kxh7 19. h5 $18) 18. Nxh7 Nxh7 (18... Kxh7?? 19. Bf4 Qd7 20. Bxb8 $18) 19. h5 Nh4 20. Bf4! { the final nail in the coffin } (20. gxh4?! Bb7 21. Bxb7 Rxb7 $18) 20... Qd8 21. gxh4! { and the rest is history } (21. Bxb8?! Nxg2 22. Kxg2 Bb7+ 23. f3 Qd5) 21... Rb7 22. h6 (22. Bxb7 Bxb7 23. f3 Bxf3 $16) 22... Qxh4 23. hxg7 (23. Bxb7? { fails to } 23... Bxb7 24. f3 Qxf4 25. hxg7 Kxg7 26. Qg2+ Ng5 $19) 23... Kxg7 { White gets strong play along the open h-file } (23... Rd8 24. Bg3 (24. Bxb7? { doesn't work because of } 24... Bxb7 25. Re4 f5 $19) 24... Qh6 25. Re3 $16 (25. Bxb7?! Bxb7 26. f3 Ng5)) 24. Re4 (24. Bxb7? { doesn't work } 24... Bxb7 25. f3 Qxf4 26. Qg2+ Ng5 $19) 24... Qh5 25. Re3 f5? (25... Rh8 26. Rh3 Qb5 27. Rg3+ Kf8 $18) 26. Rh3! { doomsday } (26. Bxb7?! { is much
worse } 26... Bxb7 27. Rg3+ Kh8 $18) 26... Qe8 27. Be5+! { a mean check } 27... Nf6 28. Qd2 { The mate threat is Qh6 } 28... Kf7 29. Qg5 { Do you see the mate threat? } (29. Bxb7?! { is the weaker alternative } 29... Bxb7 30. Bxf6 Kxf6 31. Qh6+ Ke7 $18) 29... Qe7 30. Bxf6 { Threatening mate: Rh7 } 30... Qxf6 31. Rh7+ Ke8 32. Qxf6! { the end of the
story } 32... Rxh7 (32... Rxf6 { is one last hope } 33. Bxb7 Bxb7 34. Rxb7 Rf7 35. Rxf7 Kxf7 $18) 33. Bc6+ { 1-0 Black resigns. } (33... Bd7 34. Qxe6+ Kd8 35. Qd6 Rg8+ 36. Kf1 $18) 1-0

The problem is at the end like here

 33. Bc6+ { 1-0 Black resigns. } (33... Bd7 34. Qxe6+

Bc6+ and Bd7 are in the same children of the parent, instead of Bd7 should be the children of Bc6+

**Comment by mliebelt** [comment]
The reason for that is quite clear:

* PGN itself only offers variants as variant of a main line move.
* There is no support for "a move that should be in the main line, but not part of the game".

So lichess (and others) were creative in adding the option to "enforce variant". See https://lichess.org/study/9dPKX0RR/0A6hK2CM for an example. There is no main line after move 2 of white, but 3 black moves were given as variant.

For a parser that does not know anything besides parsing moves, it is not possible to "guess" what could be meant here. So in my example, the PGN `1. e4 e5 2. Nf3 Nc6 (2... d6) (2... Nf6) *` is not the same as entered in the study, but the only way to later parse it then. (And no, I don't understand what is the difference between the study in your example and the simple PGN I have created.)

I think we have the anomaly (in lichess at least), that not all content in all studies can be saved as PGN and identically loaded into a study.

**Comment by niklasf** [comment]
Right. Work is ongoing to fix this in Lichess. First a pretty major refactoring for the internal tree representation of studies - will still take some time. And then one idea is to use null moves to represent variations like these (which is also non-standard, but should work better with all sorts of real-world PGN parsers).

**Comment by eguneys** [comment]
I don't understand what you mean, Just put them as children as intended not siblings of the parent, why would it matter it is mainline or not. It just happens to be that first children is the mainline so what.
